mediorum/server/migrate_qm.go

```diff
@@ -0,0 +1,173 @@
+	time.Sleep(time.Minute)
+
+	logger := ss.logger.With("task", "migrate_qm")
+
+	// Repeat the migration process every minute in case new uploads legacy came in (they shouldn't).
+	// Subsequent runs should be very fast since the migration deletes rows from Files after adding them to the bucket and inserting a row into the blobs table.
+	for {
+		start := time.Now()
+
+		// TODO: It would be nice if this could be an optimizely flag
+		if ss.Config.MigrateQmCids {
+			logger.Info("Qm CID migration starting")
+			err := ss.migrateQmCids()
+			took := time.Since(start)
+			if err != nil {
+				logger.Error("Qm CID migration failed", "err", err, "took", took)
+			} else {
+				logger.Info("Qm CID migration OK", "took", took)
+			}
+		} else {
+			logger.Info("Qm CID migration disabled")
+		}
+
+		logger.Info("Qm CID migration sleeping", "sleep", time.Minute)
+		time.Sleep(time.Minute)
+	}
+}
+
+func (ss *MediorumServer) migrateQmCids() error {
+	pageSize := 100
+	var lastCreatedAt time.Time
+
+	for {
+		ctx := context.Background()
+		logger := ss.logger.With("task", "migrate_qm", "lastCreatedAt", lastCreatedAt)
+
+		/* assumes unique fileUUID for each row - check with:
+		SELECT *
+		FROM "Files"
+		WHERE "fileUUID" IN (
+				SELECT "fileUUID"
+				FROM "Files"
+				GROUP BY "fileUUID"
+				HAVING COUNT("fileUUID") > 1
+		)
+		*/
+
+		// note that fileUUID doesn't have an index, so we select by createdAt and delete by fileUUID
+		baseSql := `SELECT "storagePath", "multihash", "dirMultihash", "fileName", "fileUUID", "createdAt"
+    FROM "Files"
+    WHERE "type" IN ('copy320', 'image')`
+
+		orderSql := `ORDER BY "createdAt" ASC LIMIT $1`
+
+		var rows pgx.Rows
+		var err error
+
+		if !lastCreatedAt.IsZero() {
+			sql := baseSql + ` AND "createdAt" > $2 ` + orderSql
+			rows, err = ss.pgPool.Query(ctx, sql, pageSize, lastCreatedAt)
+		} else {
+			sql := baseSql + orderSql
+			rows, err = ss.pgPool.Query(ctx, sql, pageSize)
+		}
+		if err != nil {
+			logger.Error("error querying cid storage path", "err", err)
+			continue
+		}
+
+		var fileUUIDs []string
+		var noRows bool = true
+		for rows.Next() {
+			noRows = false
+			var storagePath, multihash, dirMultihash, fileName, fileUUID string
+			var createdAt time.Time
+			err = rows.Scan(&storagePath, &multihash, &dirMultihash, &fileName, &fileUUID, &createdAt)
+			if err != nil {
+				logger.Error("error scanning row", "err", err)
+				return err
+			}
+
+			fileUUIDs = append(fileUUIDs, fileUUID)
+
+			diskPath := storagePath
+			if _, err := os.Stat(diskPath); os.IsNotExist(err) {
+				// look for the file under the legacy and custom paths (see creator-node/fsutils.ts)
+				if dirMultihash != "" {
+					diskPath = computeFilePathInDir(dirMultihash, multihash)
+				}
+				if _, err := os.Stat(diskPath); os.IsNotExist(err) {
+					diskPath = computeFilePath(multihash)
+					if _, err := os.Stat(diskPath); os.IsNotExist(err) {
+						diskPath = computeLegacyFilePath(multihash)
```

Comment:
need to `os.Stat` this, too, before continuing. probably to start out with, remove it from `fileUUIDs` so we don't delete it from the `Files` table (in case there's some other way to recover it)